File: examples/live.py
#!/usr/bin/env python3
"""
Segment live speech from the default audio device.
"""

from pocketsphinx5 import Vad, Decoder
from collections import deque
import subprocess
import sys
import os

MODELDIR = os.path.join(os.path.dirname(__file__), os.path.pardir, "model")


class Endpointer(Vad):
    def __init__(
        self,
        stream=None,
        vad_frames=10,
        vad_mode=Vad.LOOSE,
        sample_rate=Vad.DEFAULT_SAMPLE_RATE,
        frame_length=Vad.DEFAULT_FRAME_LENGTH,
    ):
        super(Endpointer, self).__init__(vad_mode, sample_rate, frame_length)
        self.buf = deque(maxlen=vad_frames)
        if stream is not None:
            self.listen(stream)

    def listen(self, stream):
        self.buf.clear()
        self.stream = stream
        self.timestamp = 0.0
        self.start_time = self.end_time = None

    def read_frame(self):
        frame = self.stream.read(self.frame_bytes)
        self.buf.append((self.is_speech(frame), self.timestamp, frame))
        self.timestamp += self.frame_length
        return sum(f[0] for f in self.buf)

    def start(self):
        while True:
            speech_count = self.read_frame()
            if speech_count > 0.9 * self.buf.maxlen:
                self.start_time = self.buf[0][1]
                self.end_time = None
                return self.start_time

    def speech(self):
        while True:
            yield self.buf.popleft()
            speech_count = self.read_frame()
            if speech_count < 0.1 * self.buf.maxlen:
                self.end_time = self.buf[-1][1]
                for frame in self.buf:
                    yield frame
                self.buf.clear()
                break


def main():
    ep = Endpointer()
    decoder = Decoder(
        hmm=os.path.join(MODELDIR, "en-us/en-us"),
        lm=os.path.join(MODELDIR, "en-us/en-us.lm.bin"),
        dict=os.path.join(MODELDIR, "en-us/cmudict-en-us.dict"),
        samprate=float(ep.sample_rate),
    )
    soxcmd = f"sox -q -r {ep.sample_rate} -c 1 -b 16 -e signed-integer -d -t raw -"
    sox = subprocess.Popen(soxcmd.split(), stdout=subprocess.PIPE)
    ep.listen(sox.stdout)
    while True:
        start_time = ep.start()
        print("Speech start at %.2f" % (start_time), file=sys.stderr)
        decoder.start_utt()
        for is_speech, timestamp, frame in ep.speech():
            decoder.process_raw(frame)
            hyp = decoder.hyp()
            if hyp is not None:
                print("PARTIAL RESULT:", hyp.hypstr, file=sys.stderr)
        print("Speech end at %.2f" % (ep.end_time), file=sys.stderr)
        decoder.end_utt()
        print(decoder.hyp().hypstr)


try:
    main()
except KeyboardInterrupt:
    pass

File: examples/live.c
/* Example of simple PocketSphinx speech segmentation.
 *
 * MIT license (c) 2022, see LICENSE for more information.
 *
 * Author: David Huggins-Daines <dhdaines@gmail.com>
 */
#include <pocketsphinx.h>
#include <signal.h>

static int global_done = 0;
static void
catch_sig(int signum)
{
    global_done = 1;
}

static FILE *
popen_sox(int sample_rate)
{
    char *soxcmd;
    size_t len;
    FILE *sox;
    #define SOXCMD "sox -q -r %d -c 1 -b 16 -e signed-integer -d -t raw -"
    len = snprintf(NULL, 0, SOXCMD, sample_rate);
    if ((soxcmd = malloc(len + 1)) == NULL)
        E_FATAL_SYSTEM("Failed to allocate string");
    if (snprintf(soxcmd, len + 1, SOXCMD, sample_rate) != len)
        E_FATAL_SYSTEM("snprintf() failed");
    if ((sox = popen(soxcmd, "r")) == NULL)
        E_FATAL_SYSTEM("Failed to popen(%s)", soxcmd);
    free(soxcmd);

    return sox;
}

int
main(int argc, char *argv[])
{
    ps_decoder_t *decoder;
    ps_endpointer_t *ep;
    cmd_ln_t *config;
    FILE *sox;
    short *frame;
    size_t frame_size;

    #ifndef MODELDIR
    #define MODELDIR "../model"
    #endif
    if ((config = cmd_ln_init(NULL, ps_args(), TRUE,
                              "-hmm", MODELDIR "/en-us/en-us",
                              "-lm", MODELDIR "/en-us/en-us.lm.bin",
                              "-dict", MODELDIR "/en-us/cmudict-en-us.dict",
                              NULL)) == NULL)
        E_FATAL("Command line parse failed\n");
    ps_default_search_args(config);
    if ((decoder = ps_init(config)) == NULL)
        E_FATAL("PocketSphinx decoder init failed\n");

    if ((ep = ps_endpointer_init(0, 0.0, 0, 0, 0)) == NULL)
        E_FATAL("PocketSphinx endpointer init failed\n");
    sox = popen_sox(ps_endpointer_sample_rate(ep));
    frame_size = ps_endpointer_frame_size(ep);
    if ((frame = malloc(frame_size * sizeof(frame[0]))) == NULL)
        E_FATAL_SYSTEM("Failed to allocate frame");
    if (signal(SIGINT, catch_sig) == SIG_ERR)
        E_FATAL_SYSTEM("Failed to set SIGINT handler");
    while (!global_done) {
        const int16 *speech;
        int prev_in_speech = ps_endpointer_in_speech(ep);
        if (fread(frame, sizeof(frame[0]), frame_size, sox) != frame_size) {
            if (!feof(sox))
                E_ERROR_SYSTEM("Failed to read %d samples", frame_size);
            /* FIXME: Should zero-pad and process, but whatever */
            break;
        }
        speech = ps_endpointer_process(ep, frame);
        if (speech != NULL) {
            const char *hyp;
            if (!prev_in_speech) {
                fprintf(stderr, "Speech start at %.2f\n",
                        ps_endpointer_segment_start(ep));
                ps_start_utt(decoder);
            }
            if (ps_process_raw(decoder, speech, frame_size, FALSE, FALSE) < 0)
                E_FATAL("ps_process_raw() failed\n");
            if ((hyp = ps_get_hyp(decoder, NULL)) != NULL)
                fprintf(stderr, "PARTIAL RESULT: %s\n", hyp);
            if (!ps_endpointer_in_speech(ep)) {
                fprintf(stderr, "Speech end at %.2f\n",
                        ps_endpointer_segment_end(ep));
                ps_end_utt(decoder);
                if ((hyp = ps_get_hyp(decoder, NULL)) != NULL)
                    printf("%s\n", hyp);
            }
        }
    }
    free(frame);
    if (pclose(sox) < 0)
        E_ERROR_SYSTEM("Failed to pclose(sox)");
    ps_endpointer_free(ep);
    ps_free(decoder);
        
    return 0;
}
